### DISCUSSION TITLE
Using ControlNet inside a custom extension

### DISCUSSION THREAD

**Discussion by volotat** [Mikubill/sd-webui-controlnet#1095] [General]
Hello. I'm trying to build an extension for creating vid2vid and txt2vid animations that heavily relies on ControlNet. I managed to hook up CN to my extension, but it behaves very strangely. First of all CN does not work until I use it inside the img2img tab, but if I run it there and then go to my tab CN suddenly starts working and even takes the right image from the right place. But if I disable CN in my tab or img2img tab it still continues to work and uses the last used image even if I delete it from the image preview. I am not well experienced with web-ui code and that is the first extension I am trying to build, so it is most likely the error from my side. Is there anyway to know what the problem might be? And is it even possible to use it inside a custom extension or was it built only to work inside txt2img and img2img tabs?

edit: It looks like any parameters I set in the CN tab in my extension will be updated into memory only after I run generation in the 'img2img' tab. The use of CN while the image is generated on the 'img2img' tab is irrelevant.

**Comment by volotat** [comment]
I was able to make it work properly by making several changes in controlnet.py script. I looked how 'txt2img_submit_button' and 'img2img_submit_button' used in the code and made references for my own generate button in Script.uigroup, Img2ImgTabTracker.on_after_component_callback and on_after_component methods. It is working correctly now, but obviously changing another extension to make mine work properly doesn't seem like a good option. Is there a better way to hook up the generate button to the CN without changing the code of the CN extension?

**Comment by ljleb** [comment]
At the moment the most straightforward way to create your own denoising pipeline based on either txt2img or img2img is to set up a script runner with the extensions you want inside, and then use `<your script runner>.run(p)` or `modules.processing.process_images(p)` as many times as you need.

Setting up a script runner can be a bit tricky, so you can also reuse the txt2img or img2img script runner, with the tradeoff of activating every other active extension using the last ui parameters of the corresponding tab.

Once you have the script runner set up, you can use the external code api to set up controlnet parameters from code. Check the wiki for this step.

Check how the img2img sequence batch tab was implemented for an example of how to launch denoising: https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/72cd27a13587c9579942577e9e3880778be195f6/modules/img2img.py#L66-L68

**Comment by volotat** [comment]
@ljleb Thank you for the response. It did helped. Everything works as intended now.
